File: datahub-console/src/main/resources/static/js/policy/naming-policy.js
/*
 * Naming policy — the browser side of the external-id naming rules.
 *
 * The platform stores external ids verbatim within a charset floor (A-Z a-z 0-9 . _ : + = -,
 * 3 to 256 characters) and layers a configurable *naming policy* on top of that. A write can
 * therefore come back three ways, and this module is the one place that knows all three:
 *
 *   - clean            : nothing to show;
 *   - allowed + warned : the response envelope carries `warnings[]` beside its items, the write
 *                        went through, and a finding was recorded for a steward;
 *   - rejected         : HTTP 400 with an RFC 9457 body whose `type` is NAMING_POLICY_ERROR and
 *                        whose `violations[]` name every offending item. Batch writes are
 *                        all-or-nothing, so the flash has to lead with "nothing was created" —
 *                        that is the first thing the person looking at it needs to know.
 *
 * It also wraps the preflight endpoint (POST /policies/naming/check), which is what lets the
 * external-id field validate as you type rather than on submit, and the findings queue — which is
 * not a findings endpoint at all: a finding is an event, so the queue reads and resolves through
 * /events like everything else in the store. See the Findings section below.
 *
 * Calls go straight to datahub-api with a bearer token from the console's /token endpoint — the
 * Feign proxy in this app is deprecated for new work (see DatahubApi's javadoc), so nothing here
 * routes through it.
 */
window.NamingPolicy = (function () {

	/** `type` on the RFC 9457 body that marks a 400 as a naming-policy rejection. */
	const NAMING_POLICY_ERROR = 'https://intellistream.ai/errors/naming-policy';

	/*
	 * The platform charset floor, mirroring TextValidator.validateExternalIdCharset. Used for the
	 * `pattern` attribute on external-id inputs so the browser blocks only what the platform would
	 * genuinely refuse. It deliberately does NOT encode a naming *convention*: which convention is
	 * in force is a per-tenant policy the browser cannot know, so that half is answered by the
	 * preflight endpoint instead of by a hardcoded regex.
	 */
	// The trailing dash is escaped deliberately. This string is used as an HTML `pattern`
	// attribute, which browsers compile with the RegExp `v` flag, and `v` mode rejects an
	// unescaped `-` at the end of a character class ("Invalid character in character class").
	// Chrome then silently DROPS the constraint, so reportValidity() started returning true
	// for any external id at all. `\-` is valid in both `u` and `v` mode and matches exactly
	// the same characters, so the floor is unchanged.
	const CHARSET_PATTERN = '[A-Za-z0-9._:+=\\-]{3,256}';

	// ---- API access ---------------------------------------------------------------------------
	// Same shape as the other browser-direct callers (home-dashboard.js, files-page.js): a short
	// token cache so a burst of keystroke-driven preflights doesn't hit /token once per call.
	let cachedToken = null;
	let cachedAt = 0;

	function apiBase() {
		const meta = document.querySelector('meta[name="datahub-api-url"]');
		return (meta ? meta.content : '').replace(/\/+$/, '');
	}

	function token() {
		if (cachedToken && (Date.now() - cachedAt) < 60000) return Promise.resolve(cachedToken);
		return fetch('/token', { headers: { Accept: 'text/plain' }, credentials: 'same-origin' })
			.then(r => r.ok ? r.text() : Promise.reject(r.status))
			.then(t => { cachedToken = t; cachedAt = Date.now(); return t; });
	}

	function apiFetch(path, opts) {
		const o = opts || {};
		return token().then(t => fetch(apiBase() + path, {
			method: o.method || 'GET',
			headers: Object.assign({
				Accept: 'application/json',
				Authorization: 'Bearer ' + t
			}, o.body ? { 'Content-Type': 'application/json' } : {}),
			body: o.body ? JSON.stringify(o.body) : undefined,
			signal: AbortSignal.timeout(10000)
		}));
	}

	// ---- Response handling --------------------------------------------------------------------

	/** True when `json` is the RFC 9457 body of a naming-policy rejection. */
	function isViolation(json) {
		return !!(json && json.type === NAMING_POLICY_ERROR);
	}

	// "<external id> — <message>", with the suggestion appended when the server offered one. The
	// suggestion is never applied automatically: not rewriting what the caller sent is the entire
	// point of this change.
	function line(externalId, message, suggestion) {
		let text = $L('policy.naming.finding.line', null, [externalId || '', message || '']);
		if (suggestion) text += '  ' + $L('policy.naming.finding.suggestion', null, [suggestion]);
		return text;
	}

	/**
	 * Flashes the warnings on a successful write, if it carries any.
	 * @return true when something was shown.
	 */
	function flashWarnings(json) {
		const warnings = json && json.warnings;
		if (!Array.isArray(warnings) || warnings.length === 0) return false;
		Flash.warning($L('policy.naming.warning.title'), {
			duration: 0,   // a steward has to read these; they should not vanish on a timer
			details: warnings.map(w => line(w.externalId, w.message, w.suggestion))
		});
		return true;
	}

	/**
	 * Flashes a naming-policy rejection, naming every offending id.
	 * @return true when `json` was a naming-policy rejection and was shown.
	 */
	function flashViolations(json) {
		if (!isViolation(json)) return false;
		const violations = Array.isArray(json.violations) ? json.violations : [];
		// The server's `detail` already states that nothing was created, and states it with the
		// real counts ("2 of 500 ..."). Prefer it, and fall back to our own wording only if a
		// server ever omits it — losing the "nothing was created" half would be the worst thing
		// to get wrong here.
		const title = json.detail || $L('policy.naming.violation.title');
		Flash.error(title, {
			duration: 0,
			details: violations.map(v => line(v.externalId, v.reason || v.message, v.suggestion))
		});
		return true;
	}

	// ---- Preflight ----------------------------------------------------------------------------

	/**
	 * Asks the API what the active policy would say about these ids, without writing anything.
	 * Findings come back only for non-OK ids, so an empty array means every id is fine.
	 *
	 * @param externalIds array of candidate ids
	 * @param dataSetId   optional; a data set may override the tenant's naming policy
	 * @return Promise<Array<{index, externalId, decision, policy, message, suggestion}>>
	 */
	function check(externalIds, dataSetId) {
		const body = { externalIds: externalIds };
		// Passed through as given, never Number()-ed. The api hands ids over as strings and reads
		// them back as strings, so the conversion buys nothing and can only lose: above 2^53 it does
		// not fail, it quietly yields a different id.
		if (dataSetId) body.dataSetId = dataSetId;
		return apiFetch('/policies/naming/check', { method: 'POST', body: body })
			.then(r => r.ok ? r.json() : Promise.reject(r.status))
			.then(json => (json && json.findings) || []);
	}

	/**
	 * Wires debounced preflight validation onto an external-id input.
	 *
	 * Validating as you type rather than on submit is the whole reason the preflight endpoint
	 * exists: a naming convention you only discover you broke after filling in a form is a
	 * convention people route around.
	 *
	 * @param input   the <input name="externalId"> element
	 * @param options {dataSetId: number|function} optional data-set context for the policy lookup
	 */
	function bindField(input, options) {
		if (!input || input.dataset.namingPolicyBound === 'true') return;
		input.dataset.namingPolicyBound = 'true';
		const opts = options || {};

		const note = Object.assign(document.createElement('div'), { className: 'naming-check' });
		input.insertAdjacentElement('afterend', note);

		let timer = null;
		let sequence = 0;   // guards against a slow early response landing after a fast later one

		const clear = () => {
			note.textContent = '';
			note.className = 'naming-check';
			input.classList.remove('naming-invalid');
		};

		const show = (state, text, suggestion) => {
			note.className = 'naming-check is-' + state;
			note.textContent = text;
			input.classList.toggle('naming-invalid', state === 'error');
			if (suggestion) {
				const apply = Object.assign(document.createElement('button'), {
					type: 'button',
					className: 'naming-check-apply',
					textContent: $L('policy.naming.check.apply')
				});
				// Offered, never applied on the user's behalf.
				apply.addEventListener('click', () => {
					input.value = suggestion;
					input.dispatchEvent(new Event('input', { bubbles: true }));
				});
				note.append(' ', apply);
			}
		};

		const run = () => {
			const value = (input.value || '').trim();
			if (!value) { clear(); return; }
			if (!new RegExp('^' + CHARSET_PATTERN + '$').test(value)) {
				// The charset floor is knowable client-side, so answer it without a round trip.
				show('error', $L('policy.naming.check.charset'));
				return;
			}
			const mine = ++sequence;
			const dataSetId = typeof opts.dataSetId === 'function' ? opts.dataSetId() : opts.dataSetId;
			check([value], dataSetId)
				.then(findings => {
					if (mine !== sequence) return;
					const finding = findings.find(f => f.externalId === value) || findings[0];
					if (!finding) { show('ok', $L('policy.naming.check.ok')); return; }
					show(finding.decision === 'NOT_OK' ? 'error' : 'warn',
						finding.message || '', finding.suggestion);
				})
				// A preflight is an aid, not a gate: if it cannot be reached the form still submits
				// and the server remains the authority.
				.catch(() => { if (mine === sequence) clear(); });
		};

		input.addEventListener('input', () => {
			clearTimeout(timer);
			timer = setTimeout(run, 400);
		});
		input.addEventListener('blur', () => { clearTimeout(timer); run(); });
	}

	// ---- Findings -----------------------------------------------------------------------------

	/**
	 * The shape of an external id: every run of uppercase letters becomes "A+", lowercase "a+",
	 * digits "9+", and separators are kept as they are. So COM-99-PT-1034 and PMP-07-XX-2201 both
	 * become "A+-9+-A+-9+".
	 *
	 * This is what the findings queue groups on, and it is the reason the queue is usable at all:
	 * a bulk import of 10,000 non-conforming resources produces 10,000 findings that are all the
	 * same mistake, and ten thousand rows is not a queue anyone works through. Grouping on the
	 * shape collapses them to one row a steward can actually judge.
	 */
	function shapeOf(value) {
		const RUNS = 'Aa9';
		let out = '';
		let previous = '';
		for (const ch of String(value == null ? '' : value)) {
			let cls;
			if (ch >= 'A' && ch <= 'Z') cls = 'A';
			else if (ch >= 'a' && ch <= 'z') cls = 'a';
			else if (ch >= '0' && ch <= '9') cls = '9';
			else cls = ch;
			if (cls === previous && RUNS.includes(cls)) continue;   // collapse the run
			out += RUNS.includes(cls) ? cls + '+' : cls;
			previous = cls;
		}
		return out;
	}

	/*
	 * A finding IS an event. There is no findings endpoint — findings are read, filtered and
	 * resolved through /events like the alarms and work orders they sit beside, which is what lets a
	 * steward see a policy complaint in the same timeline as everything else about a resource.
	 *
	 * The encoding is defined server-side in PolicyFindingEvent.java; these constants mirror it and
	 * have to be changed together with it.
	 */
	const FINDING_TYPE = 'policy_finding';
	const STATUS_OPEN = 'OPEN';
	const STATUS_RESOLVED = 'RESOLVED';

	/**
	 * Read one finding event into the flat shape the queue renders.
	 *
	 * Confining the encoding to this one function is deliberate: the rest of the page should not
	 * have to know that the policy hides in `subType` or that the offending value is a metadata key.
	 */
	function toFinding(event) {
		const metadata = (event && event.metadata) || {};
		// A list of {id, externalId} — not the bare id list the two-parallel-lists encoding used
		// before it was collapsed into one. `id` arrives as a string (the api serializes ids that way)
		// and stays one on the way back out: converting it gains nothing, and above 2^53 Number()
		// does not fail, it returns a different id.
		const related = (event && event.relatedResources) || [];
		return {
			id: event.id,
			// The correlation key: every event in this finding's lifecycle carries it, and it is what
			// the fold groups on and what a resolve names.
			externalId: event.externalId,
			nodeId: related.length && related[0].id != null ? related[0].id : null,
			dataSetId: event.dataSetId != null ? event.dataSetId : null,
			policyName: event.subType || '',
			message: event.description || '',
			suggestion: metadata.suggestion || '',
			offendingValue: metadata.offendingValue || '',
			raisedBy: metadata.raisedBy || '',
			eventTime: event.eventTime,
			resolved: event.status === STATUS_RESOLVED
		};
	}

	/**
	 * Fold a finding's events into its current state: last event wins.
	 *
	 * A finding is a stream, not a row. Raising appends OPEN, resolving appends RESOLVED with the
	 * same externalId, and a later raise for a new offending value appends another OPEN after that.
	 * Replaying in eventTime order is therefore the only way to know where a finding stands — and it
	 * is why the server cannot filter on `status` for us: a stored OPEN says "this was raised", not
	 * "this is outstanding".
	 *
	 * Events arrive already sorted ascending, so a later event for a key simply overwrites the
	 * earlier one. `dateCreated` is carried forward from the first event seen, so the queue keeps
	 * showing when the finding was first raised rather than when it was last touched — a steward
	 * sorting by age wants the age of the problem.
	 *
	 * @param states Map<externalId, finding> accumulated across pages, mutated in place
	 */
	function foldInto(states, events) {
		events.forEach(function (event) {
			const finding = toFinding(event);
			const previous = states.get(finding.externalId);
			finding.dateCreated = previous ? previous.dateCreated : finding.eventTime;
			finding.resolvedAt = finding.resolved ? finding.eventTime : null;
			// A RESOLVED event carries no message or offending value of its own — it only asserts a
			// judgement — so keep what the raise it answers said.
			if (previous) {
				finding.message = finding.message || previous.message;
				finding.offendingValue = finding.offendingValue || previous.offendingValue;
				finding.suggestion = finding.suggestion || previous.suggestion;
				finding.raisedBy = finding.raisedBy || previous.raisedBy;
				finding.nodeId = finding.nodeId != null ? finding.nodeId : previous.nodeId;
				finding.dataSetId = finding.dataSetId != null ? finding.dataSetId : previous.dataSetId;
			}
			states.set(finding.externalId, finding);
		});
		return states;
	}

	/**
	 * The page cursor for an event: `<eventTime epoch millis>_<id>`, the format the API's
	 * `cursor` field takes.
	 *
	 * The id is carried as well as the timestamp because event times are not unique — a bulk import
	 * lands thousands of findings in the same millisecond — and a cursor on the timestamp alone
	 * would either skip that group or repeat it forever.
	 */
	function cursorOf(event) {
		return String(Date.parse(event.eventTime)) + '_' + event.id;
	}

	/**
	 * POST /events/filter, narrowed to findings.
	 *
	 * Note what is NOT sent: a `status` filter. Open-ness is folded, not stored, so asking the server
	 * for OPEN events would return the raises of findings that have since been resolved. The caller
	 * folds what comes back and decides.
	 *
	 * @param params {policy, dataSetId, limit, cursor}
	 *   `cursor` is where the previous page stopped. Paging is a keyset rather than an offset
	 *   because the events store is partitioned by event time, so resuming from a timestamp skips
	 *   whole partitions instead of counting past every row ahead of it — page 400 costs what page 1
	 *   costs.
	 * @return Promise<{events, next}> — `next` is the cursor to pass back, or null at the end
	 */
	function listFindings(params) {
		const p = params || {};
		const limit = p.limit || 200;

		const filter = { type: FINDING_TYPE };
		if (p.policy) filter.subType = p.policy;
		if (p.dataSetId) filter.dataSetId = [{ id: p.dataSetId }];

		const body = {
			filter: filter,
			limit: limit,
			// Ascending is what the fold requires — replaying out of order would let a stale OPEN
			// overwrite the RESOLVED that followed it. It also has to match the order the cursor was
			// produced in, or the next page skips rows.
			sort: { property: ['eventTime'], order: 'asc' }
		};
		if (p.cursor) body.cursor = p.cursor;

		// apiFetch serializes and sets the content type; handing it a string would post a JSON
		// string literal, which the endpoint cannot bind to a filter.
		return apiFetch('/events/filter', { method: 'POST', body: body })
			.then(r => r.ok ? r.json() : Promise.reject(r.status))
			.then(json => {
				const events = (json && json.items) || [];
				const last = events.length ? events[events.length - 1] : null;
				return {
					events: events,
					// A short page is the last page. Handing back a cursor anyway would cost one
					// more round trip to discover the same thing.
					next: (last && events.length >= limit) ? cursorOf(last) : null
				};
			});
	}

	/**
	 * Resolve a finding by appending a RESOLVED event that names the same finding.
	 *
	 * Not an edit of the raise. Resolving is a judgement, not a fix — the entity still violates the
	 * policy, someone has decided that is acceptable — and recording it as its own event keeps that
	 * judgement legible next to what it answers, instead of overwriting the complaint with the verdict
	 * on it. If the id later changes to another non-conforming value the policy appends a fresh OPEN
	 * after this, and the finding is outstanding again with no reopen rule anywhere.
	 *
	 * The data set and related resource are copied from the raise deliberately: the resolve has to
	 * come back from the same filtered query the raise does, or a queue narrowed to one data set would
	 * see the complaint and miss the answer to it.
	 *
	 * @param finding a folded finding, as produced by foldInto
	 * @return Promise<'resolved'> — rejects on anything else
	 */
	function resolveFinding(finding) {
		const event = {
			externalId: finding.externalId,
			type: FINDING_TYPE,
			subType: finding.policyName,
			status: STATUS_RESOLVED,
			eventTime: new Date().toISOString(),
			relatedResources: finding.nodeId != null ? [{ id: finding.nodeId }] : []
		};
		if (finding.dataSetId != null) event.dataSetId = finding.dataSetId;

		return apiFetch('/events/create', { method: 'POST', body: { items: [event] } })
			.then(r => r.ok ? 'resolved' : Promise.reject(r.status));
	}

	/**
	 * The open findings for one node.
	 *
	 * These used to ride along free in the entity's `metadata` and deliberately no longer do:
	 * metadata is a user-editable map, so a finding kept there could be forged or silently deleted
	 * by the same caller who triggered it, which makes it a note rather than an audit record. The
	 * cost of that decision is exactly this extra fetch.
	 *
	 * An exact server-side lookup on the entity: a finding names it through `relatedResources`, which
	 * the event filter has always supported. Open-ness, though, is folded here rather than filtered —
	 * a stored OPEN says "this was raised", not "this is outstanding", so the whole stream for this
	 * entity is fetched and replayed.
	 */
	function findingsForNode(nodeId) {
		if (!nodeId) return Promise.resolve([]);
		const body = {
			filter: {
				type: FINDING_TYPE,
				relatedResources: [{ id: nodeId }]
			},
			limit: 100,
			sort: { property: ['eventTime'], order: 'asc' }
		};
		return apiFetch('/events/filter', { method: 'POST', body: body })
			.then(r => r.ok ? r.json() : Promise.reject(r.status))
			.then(json => {
				const states = foldInto(new Map(), (json && json.items) || []);
				return Array.from(states.values()).filter(f => !f.resolved);
			})
			.catch(() => []);
	}

	return {
		NAMING_POLICY_ERROR: NAMING_POLICY_ERROR,
		CHARSET_PATTERN: CHARSET_PATTERN,
		isViolation: isViolation,
		flashWarnings: flashWarnings,
		flashViolations: flashViolations,
		check: check,
		bindField: bindField,
		shapeOf: shapeOf,
		listFindings: listFindings,
		foldInto: foldInto,
		resolveFinding: resolveFinding,
		findingsForNode: findingsForNode
	};
})();

File: datahub-api-model/src/main/java/ai/intellistream/datahub/models/policy/NamingPolicy.java
// SPDX-License-Identifier: Apache-2.0
package ai.intellistream.datahub.models.policy;

import java.util.Map;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/**
 * A resolved naming policy: the preset, what a violation does, and the near-duplicate guard's mode.
 *
 * <p>Built once per resolution and cached, never per item on the write path. The regex for
 * {@link NamingPreset#PATTERN} is compiled here rather than at each evaluation for the same reason.
 *
 * @param policyId         the policy node's id, or null for {@link #shippedDefault()}
 * @param policyExternalId names the rule in errors and findings, so a caller is told which rule
 *                         rejected them rather than only that something did
 * @param preset           the convention
 * @param compiledPattern  the compiled regex when {@code preset == PATTERN}, otherwise null
 * @param mode             what a preset violation does
 * @param nearDuplicateMode what a near-duplicate does
 */
public record NamingPolicy(
        Long policyId,
        String policyExternalId,
        NamingPreset preset,
        Pattern compiledPattern,
        PolicyMode mode,
        PolicyMode nearDuplicateMode) {

    /** Metadata keys a policy node carries its naming configuration under. */
    public static final String KEY_KIND = "kind";
    public static final String KEY_PRESET = "preset";
    public static final String KEY_PATTERN = "pattern";
    public static final String KEY_MODE = "mode";
    public static final String KEY_NEAR_DUPLICATE_MODE = "nearDuplicateMode";

    /** The {@code kind} value marking a policy node as a naming policy. */
    public static final String KIND_NAMING = "naming";

    /** Name used when no policy node is configured and the shipped default applies. */
    public static final String DEFAULT_POLICY_NAME = "naming_default";

    /**
     * A cap on the supplied regex, and the reason there is one.
     *
     * <p>A caller-supplied pattern runs on the write path, so a pathological one is a way to stall
     * every ingest in the tenant. The length cap is the cheap half of the defence; the other half is
     * that matching runs against a bounded-length external id (256 characters), which keeps even a
     * badly-nested pattern's blow-up finite. Anything more (a timeout thread per match) costs more
     * than it saves at this size.
     */
    public static final int MAX_PATTERN_LENGTH = 512;

    /**
     * The number of separator-delimited segments {@link NamingPreset#QUALIFIED_TAG} requires.
     *
     * <p>Three, because that is the smallest count that can carry a qualifier, a type and a
     * sequence — the structure that makes a tag identify one thing on a site with many. Two admits
     * {@code pump-1234}, which names a class of equipment and a number rather than an asset.
     */
    public static final int MIN_QUALIFIED_SEGMENTS = 3;

    /**
     * What applies when no policy node is configured: {@code qualified_tag}, with the
     * near-duplicate guard on, and <strong>both rules warning rather than rejecting</strong>.
     *
     * <p>So the shipped default never refuses a write. Both rules are judgements the platform
     * cannot make on the caller's behalf. {@code P-101} is a legitimate ISA-5.1 loop tag at a site
     * with one pump and an under-specified name at a site with forty. {@code pump-a-01} beside an
     * existing {@code pump_a_01} is usually one asset written two ways, and occasionally two tags a
     * facility genuinely maintains apart. Nothing in a write says which, so both are recorded for a
     * steward and neither costs the caller their data.
     *
     * <p>That makes the findings queue the product of this policy, not an error path. A deployment
     * that wants either rule to block sets its own policy with {@code mode} or
     * {@code nearDuplicateMode} on reject, which is the point of the policy being configurable.
     */
    public static NamingPolicy shippedDefault() {
        return new NamingPolicy(null, DEFAULT_POLICY_NAME, NamingPreset.QUALIFIED_TAG, null,
                PolicyMode.WARN, PolicyMode.WARN);
    }

    /**
     * Read a naming policy out of a policy node's metadata map, falling back to the shipped defaults
     * for anything absent or unparseable.
     *
     * <p>Lenient by design: a policy is edited through a form and stored as loose strings, and the
     * write path is the wrong place to discover that one of them is malformed. A bad regex here
     * degrades to the preset default rather than failing every write in the tenant — and
     * {@link #validatePattern} exists so the mistake is caught at policy-save time, where it can be
     * shown to the person who made it.
     */
    public static NamingPolicy fromMetadata(Long policyId, String policyExternalId, Map<String, String> metadata) {
        Map<String, String> meta = metadata == null ? Map.of() : metadata;

        NamingPreset preset = NamingPreset.parse(meta.get(KEY_PRESET));
        Pattern compiled = null;
        if (preset == NamingPreset.PATTERN) {
            compiled = compileOrNull(meta.get(KEY_PATTERN));
            if (compiled == null) {
                // A pattern preset with no usable pattern constrains nothing; say so by falling back
                // rather than silently accepting everything under a name that implies a rule.
                preset = NamingPreset.VERBATIM_TAG;
            }
        }

        return new NamingPolicy(
                policyId,
                policyExternalId == null ? DEFAULT_POLICY_NAME : policyExternalId,
                preset,
                compiled,
                PolicyMode.parse(meta.get(KEY_MODE), PolicyMode.REJECT),
                PolicyMode.parse(meta.get(KEY_NEAR_DUPLICATE_MODE), PolicyMode.REJECT));
    }

    /**
     * Validate a supplied regex at policy-save time.
     *
     * @return null when acceptable, otherwise the reason it is not
     */
    public static String validatePattern(String pattern) {
        if (pattern == null || pattern.isBlank()) {
            return "A pattern is required when the preset is 'pattern'.";
        }
        if (pattern.length() > MAX_PATTERN_LENGTH) {
            return "Pattern must be at most " + MAX_PATTERN_LENGTH + " characters.";
        }
        try {
            Pattern.compile(pattern);
        } catch (PatternSyntaxException e) {
            return "Pattern is not a valid regular expression: " + e.getDescription();
        }
        return null;
    }

    private static Pattern compileOrNull(String pattern) {
        if (pattern == null || pattern.isBlank() || pattern.length() > MAX_PATTERN_LENGTH) {
            return null;
        }
        try {
            return Pattern.compile(pattern);
        } catch (PatternSyntaxException e) {
            return null;
        }
    }

    /** Whether {@code externalId} satisfies this policy's preset. Says nothing about near-duplicates. */
    public boolean matchesPreset(String externalId) {
        return switch (preset) {
            // The charset floor has already been enforced by validation; verbatim adds nothing.
            case VERBATIM_TAG -> true;
            case QUALIFIED_TAG -> segmentCount(externalId) >= MIN_QUALIFIED_SEGMENTS;
            case SNAKE_CASE -> SNAKE_CASE_PATTERN.matcher(externalId).matches();
            case PATTERN -> compiledPattern == null || compiledPattern.matcher(externalId).matches();
        };
    }

    /** Today's rule: lowercase alphanumeric runs joined by single underscores. */
    private static final Pattern SNAKE_CASE_PATTERN = Pattern.compile("[a-z0-9]+(?:_[a-z0-9]+)*");

    /**
     * How many separator-delimited alphanumeric runs an external id has.
     *
     * <p>A segment is a maximal run of letters and digits; everything else in the charset
     * ({@code . _ : + = -}) separates. Counting runs rather than splitting on a separator set means
     * leading, trailing and repeated separators cost nothing: {@code =K1-M3+B02} is three segments,
     * not four with an empty one, so an IEC 81346 designation is not penalised for its leading
     * {@code =}.
     *
     * <p>Written as a scan rather than a regex or {@code split} because it runs on the write path
     * for every item in every batch, and it allocates nothing.
     */
    private static int segmentCount(String externalId) {
        int segments = 0;
        boolean inSegment = false;
        for (int i = 0; i < externalId.length(); i++) {
            char c = externalId.charAt(i);
            boolean alphanumeric = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9');
            if (!alphanumeric) {
                inSegment = false;
            } else if (!inSegment) {
                segments++;
                inSegment = true;
            }
        }
        return segments;
    }

    public String describePreset() {
        return switch (preset) {
            case VERBATIM_TAG -> "verbatim_tag";
            case QUALIFIED_TAG -> "qualified_tag (at least " + MIN_QUALIFIED_SEGMENTS
                    + " separator-delimited segments)";
            case SNAKE_CASE -> "snake_case";
            case PATTERN -> "pattern " + (compiledPattern == null ? "" : "'" + compiledPattern.pattern() + "'");
        };
    }
}
